crates/nu-parser/src/parser.rs

```diff
@@ -5579,6 +5579,49 @@ pub fn parse_builtin_commands(
     }
 }
 
+fn check_record_key_or_value(
+    working_set: &StateWorkingSet,
+    expr: &Expression,
+    position: &str,
+) -> Option<ParseError> {
```

Comment:
I am wondering if we have a more general version of this already to detect illegal chars that could be conditionally extended to include the colon for records (colons should also be disallowed for variable/parameter names due to its use in the type hints)

Reply:
I could be missing something, but I don't think there is more existing code that I can use here. `check_record_key_or_value` produces an error if and only if the given Expression is a String or StringInterpolation *and* it is not a quoted string *and* contains a colon *in any of its string literal parts* (as in, it's important that interpolations like `{a: a('b:c')d}` remain allowed). This seems so inherently specific to me that I don't think there could exist much code I can reuse here. Looking at the examples you gave: colons are allowed in variable names (`let :a:b:: int = 10` just creates the variable named `$:a:b:`; the name and type are required to be whitespace-separated), and in signatures like `def [a:int:datetime:string] {}` the parser seems to just pick the name `$a` and the type `string`, ignoring valid types in between (which is.. suprising; we might want to revisit this some time?)